### ISSUE TITLE
Trait methods that return an unsized type only error when implementing trait

### ISSUE THREAD

**Issue by Aatch** [rust-lang/rust#26204]
This doesn't cause an error:

```rust
trait A { }
trait B { fn foo() -> A; }
```

despite the fact you wouldn't be able to ever implement `B`.

Making this an error should be fine for backwards compatiblity, since you'd not be able to implement the trait anyway. It's a bit confusing that the error is only when implementing the trait.

**Comment by apasel422** [comment]
Possibly related to #19182.

**Comment by arielb1** [comment]
I think this is by-design according to @nikomatsakis

**Comment by arielb1** [comment]
closed as by-design.
